tc/autotuner/utils.cc

```diff
@@ -0,0 +1,207 @@
+
+#include <algorithm>
+#include <cmath>
+#include <sstream>
+
+#include <glog/stl_logging.h>
+
+#include "tc/core/cuda/cuda_compilation_cache.h"
+#include "tc/core/cuda/cuda_mapping_options.h"
+#include "tc/core/flags.h"
+#include "tc/core/utils/dlpack.h"
+#include "tc/core/utils/time.h"
+#include "tc/lang/canonicalize.h"
+
+namespace tc {
+namespace autotune {
+
+std::vector<std::size_t> powers2andCeilDivisors(std::size_t val) {
+  auto numPowers = static_cast<std::size_t>(std::ceil(std::log2(val)));
+  // 1. generate `numPowers' powers of 2
+  std::vector<std::size_t> res(numPowers + 1);
+  std::size_t p = 1;
+  std::generate(res.begin(), res.end(), [p]() mutable {
+    auto old_p = p;
+    p *= 2;
+    return old_p;
+  });
+  // 2. additionally insert ceil(val / powers2)
+  res.reserve(res.size() * 2);
+  for (std::size_t i = 0, s = res.size(); i < s; ++i) {
+    if (res[i] > val) {
+      continue;
+    }
+    res.push_back(std::ceil(static_cast<double>(val) / res[i]));
+  }
+  std::sort(res.begin(), res.end());
+  res.erase(std::unique(res.begin(), res.end()), res.end());
+  return res;
+}
+
+struct OptionsWithMedianTime {
+  CudaMappingOptions options;
+  Duration medianRuntime;
+};
+
+std::vector<OptionsWithMedianTime> getOptionsAndMedianRuntimes(
+    const lang::CanonicalTcString& id,
+    const std::vector<const DLTensor*>& inputs,
+    const std::vector<const DLTensor*>& outputs) {
+  auto candidates =
+      OptionsCache::getCache()->retrieveOptionsAndRuntimes(id, inputs, outputs);
+
+  std::vector<OptionsWithMedianTime> c;
+  c.reserve(candidates.size());
+  std::transform(
+      candidates.begin(),
+      candidates.end(),
+      std::back_inserter(c),
+      [](const OptionsCacheRetrievalResult& rr) -> OptionsWithMedianTime {
+        return {std::move(rr.options), median(rr.recordedRuntimes)};
+      });
+  return c;
+}
+
+std::vector<CudaMappingOptions> restoreCandidates(
+    const lang::CanonicalTcString& tc,
+    const std::vector<const DLTensor*>& inputs,
+    const std::vector<const DLTensor*>& outputs) {
+  auto candidates = getOptionsAndMedianRuntimes(tc, inputs, outputs);
+  LOG_IF(INFO, candidates.size() < FLAGS_tuner_gen_restore_number)
+      << "Requested " << FLAGS_tuner_gen_restore_number
+      << " candidates but there are only " << candidates.size() << " in cache.";
+  auto restoreNumber =
+      std::min(candidates.size(), size_t(FLAGS_tuner_gen_restore_number));
+  std::sort(
+      candidates.begin(),
+      candidates.end(),
+      [](const OptionsWithMedianTime& a, const OptionsWithMedianTime& b) {
+        return a.medianRuntime < b.medianRuntime;
+      });
+  std::vector<CudaMappingOptions> res;
+  res.reserve(restoreNumber);
+  std::transform(
+      candidates.begin(),
+      candidates.begin() + restoreNumber,
+      std::back_inserter(res),
+      [](const OptionsWithMedianTime& rr) { return rr.options; });
+  return res;
+}
+
+llvm::Optional<CudaMappingOptions> getBestOptions(
```

Comment:
do we want an llvm dependence here?  (not critical)

Reply:
this will die when swap in the files, it's necessary for keeping git bisect working